32 byte pubkey nits

ajtowns opened the pull request.

sipa commented on bip-schnorr.mediawiki:
```diff
@@ -140,9 +140,9 @@ All provided signatures are valid with overwhelming probability if and only if t
 ** Let ''s<sub>i</sub> = int(sig<sub>i</sub>[32:64])''; fail if ''s<sub>i</sub> &ge; n''.
-** Let ''e<sub>i</sub> = int(hash(bytes(r) || pk<sub>i</sub> || m<sub>i</sub>)) mod n''.
+** Let ''e<sub>i</sub> = int(hash(r || pk<sub>i</sub> || m<sub>i</sub>)) mod n''.
 ** Let ''R<sub>i</sub> = lift_x(r)''; fail if ''lift_x(r)'' fails.
```
In commit "leave r as byte array":

If r is a byte array this needs to be `bytes(r)` rather than `lift_x(r)`.

sipa commented on bip-schnorr.mediawiki:
```diff
@@ -121,9 +121,9 @@ Input:
+* Let ''e = int(hash(r || pk || m)) mod n''.
 * Let ''R = sG - eP''.
 * Fail if ''infinite(R)''.
 * Fail if ''jacobi(y(R)) &ne; 1'' or ''x(R) &ne; r''.
```
In commit "leave r as byte array":

If `r` is a byte array it can't be compared with `x(R)`. I'm not sure about this change.

ajtowns replied:
Yeah, agree. Dropped the commit.

sipa commented on bip-schnorr.mediawiki:
```diff
@@ -149,22 +149,22 @@ All provided signatures are valid with overwhelming probability if and only if t
+* The secret key ''sk'': an integer in the range ''1..n-1'' chosen uniformly at random.
 
-The public key corresponding to secret key ''d'' is ''bytes(dG)''.
+The public key corresponding to secret key ''sk'' is ''bytes(skG)''.
```
In commit "secret key is sk not d'":

I'm not sure `skG` is more readable; it's easy to confuse it with `s*kG` (which is moderately sensible even).

ajtowns replied:
Yeah. I don't think `sk*G` is a great change either. But `bytes(int(sk)G)` isn't too bad, if converting sk to a byte array rather than an int directly works?

jonasnick replied:
Not really convinced by using multi character names. We use `ed` to mean multiplication. Could use `x` if you don't like `d`? ACK switching from accepting secret keys as bytes instead of ints.

jonasnick commented:
> Is it important that the secret key is a byte array?

imo better because this is the form in which most devs will get randomness (cf. random.cpp in bitcoin).

real-or-random commented:
Okay, makes sense.

sipa commented:
I prefer making everything single-character names again.

ajtowns commented:
Added a commit to name `c[1:33]` as `p` in taproot, which I think reads a little better

Undid the rearrangement, and just mentioned the we define key generation before verification in the spec introduction.

Kept calling the secret key byte array `sk`, but stopped doing math with it. Introduced `pubkey(d) = bytes(dG)` so I can say `pubkey(int(sk))` which seems okay to me, and is consistent with `pk` as the args for verification.

sipa commented on bip-schnorr.mediawiki:
```diff
@@ -111,13 +111,16 @@ The following convention is used, with constants as defined for secp256k1:
 ** The function ''point(x)'', where ''x'' is a 32-byte array, returns the point ''P = lift_x(int(x))''.
 ** The function ''hash(x)'', where ''x'' is a byte array, returns the 32-byte SHA256 hash of ''x''.
 ** The function ''jacobi(x)'', where ''x'' is an integer, returns the [https://en.wikipedia.org/wiki/Jacobi_symbol Jacobi symbol] of ''x / p''. It is equal to ''x<sup>(p-1)/2</sup> mod p'' ([https://en.wikipedia.org/wiki/Euler%27s_criterion Euler's criterion])<ref>For points ''P'' on the secp256k1 curve it holds that ''jacobi(y(P)) &ne; 0''.</ref>.
+** The function ''pubkey(d)'', where ''d'' is an integer in the range ''1..n-1'' returns ''bytes(dG)''.
```
If you define `pubkey` as a function from bytes to bytes the 2 call sites become even simpler.

sipa commented on bip-schnorr.mediawiki:
```diff
@@ -152,20 +155,22 @@ All provided signatures are valid with overwhelming probability if and only if t
+* Let ''P = d'G''
 * Let ''d = d' '' if ''jacobi(y(P)) = 1'', otherwise let ''d = n - d' ''.
-* Let ''k' = int(hash(bytes(d) || m)) mod n''<ref>Note that in general, taking the output of a hash function modulo the curve order will produce an unacceptably biased result. However, for the secp256k1 curve, the order is sufficiently close to ''2<sup>256</sup>'' that this bias is not observable (''1 - n / 2<sup>256</sup>'' is around ''1.27 * 2<sup>-128</sup>'').</ref>.
+* Let ''k' = int(hash(sk || m)) mod n''<ref>Note that in general, taking the output of a hash function modulo the curve order will produce an unacceptably biased result. However, for the secp256k1 curve, the order is sufficiently close to ''2<sup>256</sup>'' that this bias is not observable (''1 - n / 2<sup>256</sup>'' is around ''1.27 * 2<sup>-128</sup>'').</ref>.
```
This is not actually deterministic (though in a harmless way, I think), as `sk` can have two different representations for the same public key (so signing with `sk` and with `bytes(n-int(sk))` would both give signatures for the same public key, but with distinct `k`. Still, I feel using `bytes(d)` here would be more consistent (and less likely to run into weird test issues where implementations differ).

ajtowns replied:
Oh my god, how embarassment. Also good preparation for adding on zk proofs that you used a deterministic nonce for multisig stuff. Wonder if it's worth adding a footnote covering any of that?

real-or-random commented:
I prefer single char names slightly but I just noticed that `p` is already used for the field size.

ajtowns commented on bip-schnorr.mediawiki:
```diff
@@ -111,13 +111,18 @@ The following convention is used, with constants as defined for secp256k1:
 ** The function ''point(x)'', where ''x'' is a 32-byte array, returns the point ''P = lift_x(int(x))''.
 ** The function ''hash(x)'', where ''x'' is a byte array, returns the 32-byte SHA256 hash of ''x''.
 ** The function ''jacobi(x)'', where ''x'' is an integer, returns the [https://en.wikipedia.org/wiki/Jacobi_symbol Jacobi symbol] of ''x / p''. It is equal to ''x<sup>(p-1)/2</sup> mod p'' ([https://en.wikipedia.org/wiki/Euler%27s_criterion Euler's criterion])<ref>For points ''P'' on the secp256k1 curve it holds that ''jacobi(y(P)) &ne; 0''.</ref>.
+** The function ''pubkey(x)'', where ''x'' is a 32-byte array, returns ''bytes(dG)'' where ''d = int(x) mod n''.
```
(You mean the key prefixing section where it talks about signature+key malleability)

ajtowns commented:
Rebased.

Dropped the "just use pk, rather than bytes(P)"; it's consistent with bytes(r) which is easier to read than sig[0:32], implementors can figure out the shortcut if they like, and doing `bytes(point(x))` instead of `x` might help ensure your round trip isn't buggy and you don't have bugs for out-of-range inputs.

Haven't added it, but seems like `r` in the batch verification should be `r<sub>i</sub>`; it doesn't need to be remembered unlike the other vars, so it's not wrong as is, but...

> I prefer single char names slightly but I just noticed that `p` is already used for the field size.

Could change `m` to `msg` and use multi-char names for vars of type byte array (`arr` as the meta variable instead of `x` maybe), single char lowercase for ints, and single char upper case for points?

real-or-random commented:
Looks good. (I'm okay with `pk` and `sk` as used currently.)
